pyface/application.py

```diff
@@ -0,0 +1,259 @@
+        their start method, and should call the superclass start() method
+        before doing any work themselves.
+        """
+        ok = super(Application, self).start()
+        if ok:
+            # set up logging so messages can be displayed on splash screen
+            self._setup_logging()
+
+            # create the GUI so that the splash screen comes up first thing
+            self.gui
+
+        return ok
+
+    def stop(self):
+        """ Stop the application, cleanly releasing resources, if possible. """
+        ok = super(Application, self).stop()
+        if ok:
+            self._reset_logger()
+        return ok
+
+    # -------------------------------------------------------------------------
+    # 'BaseApplication' private interface
+    # -------------------------------------------------------------------------
+
+    def _run(self):
+        """ Actual implementation of running the application: starting the GUI
+        event loop.
+        """
+        # Fire a notification that the app is running.  This is guaranteed to
+        # happen after all initialization has occurred and the event loop has
+        # started.  A listener for this event is a good place to do things like
+        # opening application windows and other activities where you want the
+        # event loop running.
+        self.gui.set_trait_later(
+            self, 'application_initialized',
+            ApplicationEvent(application=self, event_type='initialized'))
+
+        # start the GUI - script blocks here
+        self.gui.start_event_loop()
+
+    # Exception handling ------------------------------------------------------
+
+    def _unhandled_exception(self, type, value, traceback):
+        """ Show an error dialog with exception information, if possible. """
+        from pyface.message_dialog import error
+
+        informative = """An unexpected error occurred: {}""".format(
+            format_exception_only(type, value))
+        detail = format_exception(type, value, traceback)
+        error("Unhandled Exception", informative=informative, detail=detail)
+
+    def _excepthook(self, type, value, traceback):
+        """ Handle any exceptions not explicitly caught
+
+        Try to display an appropriate error dialog, if possible.  Fail over to
+        the standard excepthook if that doesn't work.
+        """
+        try:
+            # try to log the exception, could fail eg. if disk full
+            logger.exception(value)
+            self._unhandled_exception(type, value, traceback)
+        except Exception:
+            # something went wrong, use base class excepthook
+            super(Application, self)._excepthook(type, value, traceback)
+        finally:
+            sys.exit(1)
+
+    # Destruction methods -----------------------------------------------------
+
+    def _can_exit(self):
+        """ Check with each window to see if it can be closed """
```

Comment:
I would add that to use this, a listener on `window.closing` has been added in `add_window`.